Add API Tests

agusirasil opened the pull request:
This new branch includes all the API tests for the Animals resource.

**Description:**
- API tests for GET, PUT, DELETE and PUSH have been added for the Animals resource.

**How to test:**
- Run `yarn test`

agusirasil committed:
API Tests Added

andy7love commented on apps/api/.github/workflows/playwright.yml:
```diff
@@ -0,0 +1,27 @@
+name: Playwright Tests
+on:
+  push:
+    branches: [ main, master ]
+  pull_request:
+    branches: [ main, master ]
```
This will enable this workflow to run on GitHub actions right away after the merge to master, we don't want that yet, you'll do that in a separate milestone.

So for the time being we can avoid running this workflow automatically by using `workflow_call`, this will also allow you to manually run it on actions if you need. 

```suggestion
on:
  workflow_call:
```

agusirasil replied:
I will make that change, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,94 @@
+  const PetList = await apiContext.get("http://localhost:8000/api/v1/animals");
+  const responseBody = await PetList.json();
+  expect(PetList.ok()).toBeTruthy();
+  console.log(responseBody);
```
`console.log` shouldn't be part of the final commit, this is used just for testing locally while coding. 

Instead of logging the response, let's add assertions on the content. 

For example check that it returns 2 animals, and one of them is called `Morro`.

agusirasil replied:
Done, thanks!

agusirasil replied:
Currently comparing the result of this GET request with the theoretical expected result.

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,94 @@
+    },
+  }
+});
+
```
Can you wrap all your tests in a single `describe`? 

It will help to distinguish the whole tests suite for the Animals resource. 

Also, this will let you add `beforeEach`, and `afterAll`, accordingly to prepare for the tests (re-use the apiContext, and API base URL),  and to clean-up after the tests.

agusirasil replied:
Done, thanks!

agusirasil replied:
Added the apiContext to a first test which is triggered beforeAll.

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,94 @@
+
+test("Get pet list", async () => {
+  const apiContext = await request.newContext();
+  const PetList = await apiContext.get("http://localhost:8000/api/v1/animals");
```
Do we have a way to configure the API base url (I mean `http://localhost:8000/api/v1/`) so we don't need to add it on each test?

agusirasil replied:
Done. Added it to a first test which is triggered beforeAll. 
Thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,94 @@
+  });
+  expect(newPet.ok()).toBeTruthy();
+  const requestStatus = await newPet.status();
+  console.log(requestStatus);
```
Same here, can we add more assertions ensuring the API actually added the new pet correctly with the data we introduced?

agusirasil replied:
Done. I am currently comparing the obtained data from the POST request with the data received from the GET request.
Thanks!

agusirasil replied:
Already done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,94 @@
+  const firstPet = await apiContext.get("http://localhost:8000/api/v1/animals/1");
+  const responseBody = await firstPet.json();
+  expect(firstPet.ok()).toBeTruthy();
+  console.log(responseBody);
```
Same here, remove the `console.log` and instead let's add more assertions ensuring we get enough data from this `responseBody`.

agusirasil replied:
Done, thanks!

agusirasil replied:
Also, currently comparing the data obtained in the GET request with the original set of data obtained in the GET pet list request.
Thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,94 @@
+      }
+    }
+  });
+  expect(updatedPet.ok()).toBeTruthy();
+  const requestStatus = await updatedPet.status();
+  console.log(requestStatus);
```
Same as with the add pet tests, here you're only checking that the request returned ok, then doing a `console.log` of the status. But yet you're not doing assertions on the returned data, nor if the API actually modified the 2nd resource to be named Toto. 

Maybe a GET to `animals/2` is needed to check if the new data retrieved from the 2nd resource actually matches the new data.

agusirasil replied:
Currently comparing the PUT request result with the GET result.
Thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,94 @@
+  const apiContext = await request.newContext();
+  const deletedPet = await apiContext.delete("http://localhost:8000/api/v1/animals/3");
+  expect(deletedPet.ok()).toBeTruthy();
+  const requestStatus = await deletedPet.status();
+  console.log(requestStatus);
```
Same here, so far you're only checking that the API responded ok, but still we need to assert if this resource with id 3 was actually removed.

agusirasil replied:
Done. I added an assertion.
Thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,94 @@
+  console.log(requestStatus);
+});
+
+test("Get spec", async () => {
+  const apiContext = await request.newContext();
+  const specResponse = await apiContext.get("http://localhost:8000/api/v1/spec");
+  const requestStatus = await specResponse.status();
+  expect(specResponse.ok()).toBeTruthy();
+  console.log(requestStatus);
+  console.log(specResponse.headers());
```
Get spec is not part of the **Animals** resource, should be removed from this file.

agusirasil replied:
Done, thanks!

andy7love reviewed (changes requested):
Aside from the comments below, please note that the tests added are only "happy path", the API should also be tested for non-happy paths like: 
- What if we do a GET, PUT, or DELETE on a non-existing resource? 404?
- What if the list is empty? 
- What if we do add or edit with invalid Pet data? (Wrong parameters)

agusirasil committed:
Update apps/api/.github/workflows/playwright.yml

Co-authored-by: Andres Amor <andres.amor@toptal.com>

agusirasil commented:
Thank you for your reviews. I corrected and added new tests as the happy paths were the first ones I attempted to create. I hope it now looks more complete.
If there are any new comments please let me know.

agusirasil committed:
Corrections and new tests

agusirasil committed:
small fix, removes a few repetitive lines

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+export default defineConfig({
+  use: {
+    proxy: {
+      server: 'http://local:8080',
```
I'm wondering, why you're defining the proxy server here and also setting the whole URL on line 17. 
Is this required at all? Or can we re-use this to set the whole "base url" so we don't need `url` later?

agusirasil replied:
Done, deleted the proxy server configuration and created a variable for the url. Thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+test.describe('Animals Resource', () => {
+  test.describe.configure({ mode: 'serial' });
+
+  //Preparation Test
```
```suggestion
```
This comment is not necessary, also what's on `beforeAll` is not a test, just fixture preparation.

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+import NewPetDataSetFailWrongParameter from '../utils/NewPetDataSetFailWrongParameter.json'
+import NewPetDataUpdate from '../utils/NewPetDataSetUpdate.json'
+
+let url: string;
```
```suggestion
const apiUrl: string = "http://localhost:8000/api/v1";
```
This can be a const since it's a fixed string. And you can remove the assignment on line 26. 

Also, WDYT about changing the name to be `apiUrl`? so it's clearer what URL is about?

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+    },
+  }
+});
+import AnimalsListDataSet from '../utils/AnimalsListTestsData.json'
+import NewPetData from '../utils/NewPetDataSet.json'
+import NewPetDataFail from '../utils/NewPetDataSetFail.json'
+import NewPetDataFailEmpty from '../utils/NewPetDataFailEmpty.json'
+import NewPetDataSetFailWrongParameter from '../utils/NewPetDataSetFailWrongParameter.json'
+import NewPetDataUpdate from '../utils/NewPetDataSetUpdate.json'
```
WDYT on moving all these jsons inside `tests` folder under a new folder `data ` ? 
So it will be clear that all these files' purpose is for the test, it is clear that's test data, and so we don't need to put TestData or DataSet as suffix for the filenames. 

So in the end the file could be `/apps/api/tests/data/animals-list.json`. 

Also, WDYT on sticking to the convention to use [kebab-case](https://developer.mozilla.org/en-US/docs/Glossary/Kebab_case) for the json filenames?

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+  });
+
+  //POST Add New Pet Pass
+  test("POST usign the correct data should add a new pet", async () => {
```
```suggestion
  test("POST using the correct data should add a new pet", async () => {
```

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+  });
+
+  test("GET animal with a non-existent id should return 404", async () => {
+    const firstPet = await apiContext.get(`${url}/animals/45`);
```
`firstPet` doesn't seems to be a good variable name for this case.

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+    expect(postPetRequest.status()).toBe(400);
+  });
+
+  test("POST usign the incorrect resource should return 404", async () => {
```
```suggestion
  test("POST using the incorrect resource should return 404", async () => {
```

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+    expect(postPetRequest.status()).toBe(400);
+  });
+
+  test("POST usign invalid parameters should return 400", async () => {
```
```suggestion
  test("POST using invalid parameters should return 400", async () => {
```

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+  });
+
+  //POST Add New Pet Fails
+  test("POST usign the incorrect data should return 400", async () => {
```
```suggestion
  test("POST using the incorrect data should return 400", async () => {
```

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+    expect(postPetRequest.status()).toBe(404);
+  });
+
+  test("POST usign an empty data set should return 400", async () => {
```
```suggestion
  test("POST using an empty data set should return 400", async () => {
```

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+
+  //PUT (Update) Pets Pass
+  test("PUT using the correct data should update a pet", async () => {
+    const PutUpdatedPet = await apiContext.put(`${url}/animals/2`,
```
```suggestion
    const putUpdatedPet = await apiContext.put(`${url}/animals/2`,
```
Try to keep consistency on the variable names convention. We are using [lowerCamerCase](https://wiki.c2.com/?LowerCamelCase) for this project.

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+      {
+        data: NewPetDataUpdate
+      });
+      expect(PutUpdatedPet.status()).toBe(404);
```
Indentation here is weird, maybe a running "format document" will fix it.

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+  test("PUT using the incorrect data should return 400", async () => {
+    const PutUpdatedPet = await apiContext.put(`${url}/animals/2`,
+      {
+        data: NewPetDataFail
```
```suggestion
        data: NewPetWithIncorrectData
```
Maybe we could change the name of the file and variable to be more descriptive of the content.

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+  test("PUT using invalid parameters should return 400", async () => {
+    const PutUpdatedPet = await apiContext.put(`${url}/animals/2`,
+      {
+        data: NewPetDataSetFailWrongParameter
```
```suggestion
        data: NewPetWithInvalidParameter
```
Same as above, just a slight name change of the file and variable name to keep it more consistent as the previous file but also descriptive of the content.

agusirasil replied:
Done, thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+    expect(PutUpdatedPet.status()).toBe(400);
+  });
+
+  //DELETE Pets Pass
```
```suggestion
```
Not needed.

agusirasil replied:
This comments help me read the code better, they can be ignored so I believe it won't be an issue to leave them. Thanks!

andy7love commented on apps/api/tests/Animals.spec.ts:
```diff
@@ -0,0 +1,201 @@
+    expect(deletedPet).toBeOK();
+    
+    //GET Request should be 404
+    const getNewPetResponse = await apiContext.get(`${url}/animals/2`);
```
I don't think `getNewPetResponse` is a good variable name for this get request.

agusirasil replied:
Changed to getPetResponse. Thanks!

andy7love reviewed (commented):
Overall looks good, great job! 

Still, I'd like to suggest some minor changes to improve code consistency, readability, and maintainability, 

Aside from the comments below, one more thing to think about: 
Create your own convention on how to easily name variables for these tests, and then stick to it everywhere. Right now it is not consistent, for example: 
- Sometimes **Requests** are named without suffix, and with [PascalCase](https://wiki.c2.com/?PascalCase): 
```javascript
const PetList = await apiContext.get(`${url}/animals`);
```
- But sometimes with `Request` suffix and [lowerCamelCase](https://wiki.c2.com/?LowerCamelCase): 
```javascript
const postPetRequest = await apiContext.post(`${url}/animals`, {
      data: NewPetData
});
```
- Also sometimes the **Response data** sometimes is called with `RequestData` suffix: 
```javascript
const postPetRequestData = await postPetRequest.json();
```
- And sometimes just `resopnseBody`: 
```javascript
const responseBody = await PetList.json();
```

> There are only two hard things in Computer Science: cache invalidation and naming things.
-- Phil Karlton

agusirasil commented:
I read the comments and made the neccessary changes. Took the time to rename all variables and files. Thank you!

agusirasil committed:
Changed variable and file names

agusirasil committed:
changes the name of a variable

agusirasil committed:
reviewed variables names

andy7love reviewed (approved):
Great job! 👏 👏 👏
